### PR TITLE
Install from GitHub releases!

### PR DESCRIPTION
**Pull request by fmota-da**
Replaces the bintray backend with a github backend for `daml install`.

### CODE REVIEW

File: daml-assistant/src/DAML/Assistant/Install/Github.hs [hunk 1]
```diff
@@ -0,0 +1,112 @@
+-- Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | Discover releases from the digital-asset/daml github.
+module DAML.Assistant.Install.Github
```
Comment (neil-da) [inline daml-assistant/src/DAML/Assistant/Install/Github.hs]: Export list please

File: daml-assistant/src/DAML/Assistant/Install/Github.hs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+import Network.HTTP.Simple
+import Control.Exception.Safe
+import Control.Monad
+import Data.List
+import Data.Either.Extra
+import qualified System.Info
+import qualified Data.Text as T
+
+newtype Tag = Tag Text deriving (Eq, Show, FromJSON)
+newtype AssetName = AssetName { unAssetName :: Text } deriving (Eq, Show, FromJSON)
```
Comment (neil-da) [inline daml-assistant/src/DAML/Assistant/Install/Github.hs]: Can we have an example asset name as documentation? e.g. is Gold a valid asset?

File: daml-assistant/src/DAML/Assistant/Install/Github.hs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import Data.Aeson
+import Network.HTTP.Simple
+import Control.Exception.Safe
+import Control.Monad
+import Data.List
+import Data.Either.Extra
+import qualified System.Info
+import qualified Data.Text as T
+
+newtype Tag = Tag Text deriving (Eq, Show, FromJSON)
```
Comment (neil-da) [inline daml-assistant/src/DAML/Assistant/Install/Github.hs]: Can we have an example tag name as document. Is it v1.0, 1.0, or is tag something totally different?

File: daml-assistant/src/DAML/Assistant/Install/Github.hs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    , assetDownloadURL :: InstallURL
+    } deriving (Eq, Show)
+
+instance FromJSON Asset where
+    parseJSON = withObject "Asset" $ \r ->
+        Asset
+        <$> r .: "name"
+        <*> r .: "browser_download_url"
+
+versionToTag :: SdkVersion -> Tag
```
Comment (neil-da) [inline daml-assistant/src/DAML/Assistant/Install/Github.hs]: If Tag must be isomorphic to SdkVersion, why define Tag at all? Why not just pass around SdkVersion?

File: daml-assistant/src/DAML/Assistant/Install/Github.hs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    mapLeft (assistantErrorBecause ("Tag " <> t <> "does not represent a valid SDK version.")) $
+        if T.take 1 t == "v" then
+            mapLeft (pack . displayException) $ parseVersion (T.drop 1 t)
+        else
+            Left "Tag must start with v followed by semantic version."
+
+
+makeAPIRequest :: FromJSON t => Text -> IO t
+makeAPIRequest path = do
+    request <- parseRequest (unpack ("GET https://api.github.com/repos/digital-asset/daml" <> path))
```
Comment (neil-da) [inline daml-assistant/src/DAML/Assistant/Install/Github.hs]: GitHub has an API limit. How many simultaneous SDK users do we need to break it? Note that the HLint CI script has been known to break the GitHub API limit and thus start failing, so it isn't that high. If we're as popular as a relatively recent and less common feature of HLint we'll fail. I suspect this is an unworkable approach... (as in you'd have to mirror it, cache it, create a github token or something)

File: daml-assistant/src/DAML/Assistant/Types.hs [hunk 6]
```diff
@@ -68,6 +69,10 @@ data InstallOptions = InstallOptions
     , iQuiet :: QuietInstall
     } deriving (Eq, Show)
 
+newtype InstallURL = InstallURL
```
Comment (neil-da) [inline daml-assistant/src/DAML/Assistant/Types.hs]: Example of what an install url looks like please. Does it have .tar.gz at the end? .zip? Qualified? With https?

File: daml-assistant/src/DAML/Assistant/Install/Github.hs (hunk 5 above)
Comment (moritzkiefer-da) [reply daml-assistant/src/DAML/Assistant/Install/Github.hs]: Repeating what we discussed on slack: Github has a limit of 60 requests per hour per IP for unauthenticated requests. An individual user is unlikely to hit this but we risk hitting this when IPs are shared, e.g., on Travis or if lots of people sit behind the same NAT. For now, it is probably fine. It should be easy to setup a mirror if we run into this.

File: daml-assistant/src/DAML/Assistant/Install/Github.hs (hunk 4 above)
Comment (fmota-da) [reply daml-assistant/src/DAML/Assistant/Install/Github.hs]: They aren't isomorphic. Tag represents git tags, only some of which represent valid versions.

**Comment by garyverhaegen-da** [comment]
> Approving after discussions about API limits. Would be good if some of that discussion landed somewhere in the code.

I think a good way to do that would be to explicitly handle the 403 response we would get when API limits are exceeded and:
- Tell the user about it; the explanation to the user would summarize the discussion, and
- Tell the telemetry server about it; this is not so different from an error in the IDE, and we would probably want to know how often this happens.

We could also have the assistant itself throttle its requests by caching the results and only going to GitHub once every 5 minutes or something.

File: daml-assistant/src/DAML/Assistant/Install/Github.hs (hunk 1 above)
Comment (fmota-da) [reply daml-assistant/src/DAML/Assistant/Install/Github.hs]: :+1:

File: daml-assistant/src/DAML/Assistant/Install/Github.hs (hunk 2 above)
Comment (fmota-da) [reply daml-assistant/src/DAML/Assistant/Install/Github.hs]: :+1:

File: daml-assistant/src/DAML/Assistant/Install/Github.hs (hunk 3 above)
Comment (fmota-da) [reply daml-assistant/src/DAML/Assistant/Install/Github.hs]: :+1:

File: daml-assistant/src/DAML/Assistant/Types.hs (hunk 6 above)
Comment (fmota-da) [reply daml-assistant/src/DAML/Assistant/Types.hs]: :+1:

**Comment by fmota-da** [comment]
> > Approving after discussions about API limits. Would be good if some of that discussion landed somewhere in the code.
> 
> I think a good way to do that would be to explicitly handle the 403 response we would get when API limits are exceeded and:
> 
> * Tell the user about it; the explanation to the user would summarize the discussion, and

Agreed.

> * Tell the telemetry server about it; this is not so different from an error in the IDE, and we would probably want to know how often this happens.

At a later date. Not sure what the best way to tie the assistant into telemetry right now -- the telemetry is ide only at the moment.

> We could also have the assistant itself throttle its requests by caching the results and only going to GitHub once every 5 minutes or something.

We could, but this API call is something that you only need to do once per SDK version / once per project.

**Pull request merged**